### PR TITLE
Added functionality for thermal-chamber testing experiment.

### PR DESCRIPTION
**Pull request by Siegfriedchao**
Due to the limited space, the following operations are performed:

1) Extensive `addition` and `multiplication` operations to stress the `KL03` microcontroller.

2) An array of `uint8_t` is used (size of 184), with bit-wise toggling operation to verify checksums.

3) Extensive I2C read/write operations to test both the `KL03` microcontroller and `MMA8451Q` accelerometer.

4) On-board LED is used to indicate working condition.

5) Note: Need a forced quit when exiting the stress test mode, ideally it should be run as long as possible.

### CODE REVIEW

**Comment by Siegfriedchao** [comment]
Changes made:
(1) Added a menu option to change I2C addresses
(2) Minor fix for error detection and simple multiplication and addtion operations
(3) There are still 3 GPIO pins settings (for R,G,B LEDs on KL03 board) being board specific, alternatively to add another menu option to manually set GPIOs interactively, but this is not at present a high-priority work.

**Comment by Siegfriedchao** [comment]
Changes made:
(1) Removed duplicate `read()` function.
(2) Added `write()` function in `devMMA8451Q.c` without having to copy codes in `menu` `case 'f'` for continuous write to MMA8451Q via I2C.
(3) Added printouts to indicate success/failure for writing and reading via I2C.
(4) Further code/comments cleanup.

File: src/boot/ksdk1.1.0/warp-kl03-ksdk1.1-boot.c [hunk 1]
```diff
@@ -1654,6 +1807,291 @@ main(void)
...
+										1);
+
+				uint8_t accValue[3] = {0,0,0};
+
+				/*	
+				 *	Fill up the remaining memory space using a fixed size array
+				 *	The size of this array is highly dependent on the firmware code size
+				 */
+
+				int NUMBER = 200;
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/warp-kl03-ksdk1.1-boot.c]: Still magic numbers. These should be in a config enum as we've discussed a few times now.

File: src/boot/ksdk1.1.0/warp-kl03-ksdk1.1-boot.c [hunk 2]
```diff
@@ -1654,6 +1807,291 @@ main(void)
...
+				OSA_TimeDelay(1000);
+				GPIO_DRV_ClearPinOutput(kBlue);
+				OSA_TimeDelay(1000);
+				SEGGER_RTT_printf(0, "\n\nFilling memory");
+
+				for (uint16_t i = 0; i < NUMBER; i++)
+				{
+					if(i % 2 == 0)
+					{
+						memoryPointer[i] = 0b00110011;
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/warp-kl03-ksdk1.1-boot.c]: Magic numbers

File: src/boot/ksdk1.1.0/warp-kl03-ksdk1.1-boot.c [hunk 3]
```diff
@@ -1654,6 +1807,291 @@ main(void)
...
+					}
+					if(i % 2 != 0)
+					{
+						memoryPointer[i] = 0b11001100;
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/warp-kl03-ksdk1.1-boot.c]: Magic numbers

File: src/boot/ksdk1.1.0/warp-kl03-ksdk1.1-boot.c [hunk 4]
```diff
@@ -1654,6 +1807,291 @@ main(void)
...
+					 *	(2) bit-wise toggling and checksum operations are performed
+					 *	(3) checking the I2C read and write status
+					 */
+
+					GPIO_DRV_ClearPinOutput(kGreen);
+					GPIO_DRV_SetPinOutput(kRed);
+					GPIO_DRV_SetPinOutput(kBlue);
+					OSA_TimeDelay(100);
+
+					endlessLoop(65535);
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/warp-kl03-ksdk1.1-boot.c]: Magic numbers

File: src/boot/ksdk1.1.0/warp-kl03-ksdk1.1-boot.c [hunk 5]
```diff
@@ -1965,6 +2127,282 @@ main(void)
...
+				 */
+
+				WarpStatus i2cWriteStatusA, i2cWriteStatusB;
+				WarpStatus i2cReadStatusX = kWarpStatusDeviceCommunicationFailed;
+				WarpStatus i2cReadStatusY = kWarpStatusDeviceCommunicationFailed;
+				WarpStatus i2cReadStatusZ = kWarpStatusDeviceCommunicationFailed;
+
+				enableI2Cpins(menuI2cPullupValue);
+
+				i2cWriteStatusA = writeSensorRegisterMMA8451Q(0x09 /* register address F_SETUP */,
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/warp-kl03-ksdk1.1-boot.c]: You don't check the status.

File: src/boot/ksdk1.1.0/warp-kl03-ksdk1.1-boot.c [hunk 6]
```diff
@@ -1965,6 +2127,282 @@ main(void)
...
+										0x00 /* payload: Disable FIFO */,
+										1);
+
+				i2cWriteStatusB = writeSensorRegisterMMA8451Q(0x2A /* register address */,
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/warp-kl03-ksdk1.1-boot.c]: You don't check the return status.

File: src/boot/ksdk1.1.0/gpio_pins.h [hunk 7]
```diff
@@ -3,7 +3,7 @@
 
 #include "fsl_gpio_driver.h"
 
-
+#define WARP_BUILD_ENABLE_THERMALCHAMBERANALYSIS
```
[the remark this answers is not in the document]
Comment (phillipstanleymarbell) [reply src/boot/ksdk1.1.0/gpio_pins.h]: Any code that includes `gpis_pins.h` will thus define `WARP_BUILD_ENABLE_THERMALCHAMBERANALYSIS` which will in turn enable all the code guarded by `WARP_BUILD_ENABLE_THERMALCHAMBERANALYSIS`.

File: src/boot/ksdk1.1.0/warp-kl03-ksdk1.1-boot.c [hunk 8]
```diff
@@ -1642,6 +1799,138 @@ main(void)
 				break;
 			}
 
+			case 'u':
```
[the remark this answers is not in the document]
Comment (phillipstanleymarbell) [reply src/boot/ksdk1.1.0/warp-kl03-ksdk1.1-boot.c]: This is not resolved. There is already a switch on `menuTargetSensor` and the code below duplicates that. When `menuTargetSensor` is set, you can set a pointer to the corresponding struct so you don't need to repeat this whole switch here.

File: src/boot/ksdk1.1.0/gpio_pins.c [hunk 9]
```diff
@@ -179,6 +181,26 @@ gpio_output_pin_user_config_t	outputPins[] = {
 		.config.slewRate = kPortSlowSlewRate,
 		.config.driveStrength = kPortLowDriveStrength,
 	},
+#endif
```
[the remark this answers is not in the document]
Comment (Siegfriedchao) [reply src/boot/ksdk1.1.0/gpio_pins.c]: It's part of the `#ifdef WARP_FRDMKL03` from line 147, another #ifdef then follows in line 185 which serves for chamber testing. Both of these two blocks are possible to enable at the same time, therefore not using #ifdef #elseifdef #else ... instead kept them as two separate blocks.

**Comment by Siegfriedchao** [comment]
Is `menuTargetSensor` still an unresolved problem?

File: src/boot/ksdk1.1.0/gpio_pins.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (Siegfriedchao) [reply src/boot/ksdk1.1.0/gpio_pins.c]: Checked. No errors or warnings from this part, both defined and not defined. 
As for the general build, no errors or warnings when compiling as well. One minor issue: the programme would not process properly if we are defining WARP_BUILD_ENABLE_THERMALCHAMBERANALYSIS without removing one or two of the sensor drivers (i.e. to undef one or two drivers). Reducing the 'magic' number for size definition of the array buffer `memoryFillingBuffer` would solve this problem. However, considering that we have the flexibility of undefining sensor drivers, this tends to be a minor issue.

File: src/boot/ksdk1.1.0/warp-kl03-ksdk1.1-boot.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (phillipstanleymarbell) [reply src/boot/ksdk1.1.0/warp-kl03-ksdk1.1-boot.c]: Under the top-level `a` menu item, when `menuTargetSensor` is set, you can also set the target I2C address. That will remove the need for the ~200 lines of code below.

File: src/boot/ksdk1.1.0/warp-kl03-ksdk1.1-boot.c [hunk 10]
```diff
@@ -1965,6 +2126,303 @@ main(void)
...
+				WarpStatus i2cReadStatusY = kWarpStatusDeviceCommunicationFailed;
+				WarpStatus i2cReadStatusZ = kWarpStatusDeviceCommunicationFailed;
+
+				enableI2Cpins(menuI2cPullupValue);
+
+				i2cWriteStatusA = writeSensorRegisterMMA8451Q(0x09 /* register address F_SETUP */,
+										0x00 /* payload: Disable FIFO */,
+										1);
+
+				i2cWriteStatusB = writeSensorRegisterMMA8451Q(0x2A /* register address */,
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/warp-kl03-ksdk1.1-boot.c]: Still not checking i2cWriteStatusB

**Pull request merged**